### DISCUSSION TITLE
ssh autocomplete -- where do the hosts come from?

### DISCUSSION THREAD

**Discussion by clarkritchie** [marlonrichert/zsh-autocomplete#295] [Q&A] [answered]
When I type `ssh` the list of hosts that are available for me to autocomplete includes some very old systems I once used, some as far back as 10 years ago!  I've had numerous new computers since then and simply cannot reason how how those are displayed.  I'm on OS/X and have checked my Keychain thinking that somehow they came from there via an old iCloud account maybe but I can't find them anywhere.  Any thoughts?  I'd like to purge them.  Thanks!

**Comment by marlonrichert** [comment, marked as answer]
On macOS, [the `_hosts` function of Zsh's completion system](https://github.com/zsh-users/zsh/blob/zsh-5.8/Completion/Unix/Type/_hosts) gets hosts
* [from `/etc/hosts`](https://github.com/zsh-users/zsh/blob/77d203f3fbbd76386bf197f9776269a1de580bb5/Completion/Unix/Type/_hosts#L24) and 
* [from `~/.ssh/known_hosts`](https://github.com/zsh-users/zsh/blob/77d203f3fbbd76386bf197f9776269a1de580bb5/Completion/Unix/Type/_hosts#L39). 

You can override this behavior 
* [with `zstyle ':completion:*:hosts' hosts <hostname> ...`](https://github.com/zsh-users/zsh/blob/77d203f3fbbd76386bf197f9776269a1de580bb5/Completion/Unix/Type/_hosts#L7), which overrides all of the above, or 
* [with `zstyle ':completion:*:hosts' known-hosts-files <filepath> ...`](https://github.com/zsh-users/zsh/blob/77d203f3fbbd76386bf197f9776269a1de580bb5/Completion/Unix/Type/_hosts#L39), which overrides only the second point above.
